src/wx/widgets/sdljoy.cpp

```diff
@@ -38,420 +39,472 @@ static int16_t axisval(int16_t x)
-
-                        joy.second.button[but] = state;
-
-                        wxLogDebug("POLLED SDL_JOYBUTTON: joy:%d but:%d val:%d prev_val:%d", joy.first, but, state, last_state);
-                    }
-                }
-
-                for (uint8_t axis = 0; axis < SDL_JoystickNumAxes(joy.second.dev); axis++) {
-                    auto val      = axisval(SDL_JoystickGetAxis(joy.second.dev, axis));
-                    auto prev_val = joy.second.axis[axis];
 
-                    if (val != prev_val) {
-                        CreateAndSendEvent(joy.first, WXSDLJOY_AXIS, axis, val, prev_val);
-
-                        joy.second.axis[axis] = val;
-
-                        wxLogDebug("POLLED SDL_JOYAXISMOTION: joy:%d axis:%d val:%d prev_val:%d", joy.first, axis, val, prev_val);
-                    }
-                }
-            }
+            got_event = true;
+            break;
         }
-    }
-}
 
-void wxSDLJoy::ConnectController(uint8_t joy)
-{
-    SDL_Joystick* js_dev = nullptr;
-    bool is_gc;
-
-    if ((is_gc = SDL_IsGameController(joy))) {
-        auto dev = SDL_GameControllerOpen(joy);
+        case SDL_JOYHATMOTION:
+        {
+            SDL_JoystickID joy_id = e.jhat.which;
+            auto joy_state = joystick_states_.find(joy_id);
+            if (joy_state != joystick_states_.end()) {
+                // Do not process joystick events for game controllers.
+                if (joy_state->second->IsGameController())
+                    break;
 
-        if (dev) {
-            joystate[joy].dev = dev;
+                joy_state->second->ProcessEvent(
+                    e.type, e.jhat.hat, AxisValueToDirection(e.jhat.value));
+            }
 
-            js_dev = SDL_GameControllerGetJoystick(dev);
+            got_event = true;
+            break;
         }
-    }
-    else {
-        if ((js_dev = SDL_JoystickOpen(joy)))
-            joystate[joy].dev = js_dev;
-    }
-
-    if (js_dev) {
-        auto instance = SDL_JoystickInstanceID(js_dev);
-
-        instance_map[instance] = &(joystate[joy]);
-
-        joystate[joy].instance = instance;
-    }
-
-    joystate[joy].index = joy;
-    joystate[joy].is_gc = is_gc;
-}
-
-void wxSDLJoy::RemapControllers()
-{
-    for (auto&& joy : joystate) {
-        auto& state = joy.second;
-
-        DisconnectController(state);
-        ConnectController(joy.first);
-    }
-}
 
-void wxSDLJoy::DisconnectController(wxSDLJoyState& state)
-{
-    if (auto& dev = state.dev) {
-        if (state.is_gc) {
-            SDL_GameControllerClose(dev);
+        case SDL_JOYDEVICEADDED:
+        {
+            int sdl_index = e.jdevice.which;
+            // Always remap all controllers.
+            RemapControllers();
+            got_event = true;
+            break;
         }
-        else {
-            SDL_JoystickClose(dev);
+
+        case SDL_JOYDEVICEREMOVED:
+        {
+            SDL_JoystickID joy_id = e.jdevice.which;
+            if (poll_all_joysticks_)
```

Comment:
Can this be removed? That is, can we remove the state in the `poll_all_joysticks_` case too, or would it cause issues in the input config dialog? This is my only remaining issue with the code as it is.

Reply:
I thought about it and the only issue we would have is that we may end up with the wrong player index if a controller gets disconnected while we are in the input config dialog. But in the end, I think that it is very unlikely and if it does happen, all the user has to do is either reconnect the other joystick or re-open the configuration dialog, which is a very minor annoyance. I think most users will intuit that the player index depends on how many controllers they have connected and that it might change if they have more than one.

The `poll_all_joysticks_` variable also serves another purpose, which is to differentiate the polling mode (some joystick vs. all joysticks). To work around that, I have switched to another variable `is_polling_active_`, which is set to true when either mode is active.